File: src/ros2_bridge/mappers/geometry_msgs/pose_with_covariance.rs
//! Typed mapper for `geometry_msgs/msg/PoseWithCovariance`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn pose_with_covariance_to_bus(msg: ros_env::geometry_msgs::msg::PoseWithCovariance) -> crate::geometry_msgs::msg::v1::PoseWithCovariance {
    crate::geometry_msgs::msg::v1::PoseWithCovariance {
        pose: Some(crate::ros2_bridge::mappers::geometry_msgs::pose::pose_to_bus(msg.pose)),
        covariance: crate::ros2_bridge::mappers::convert::f64_seq(msg.covariance),
    }
}

pub(crate) fn pose_with_covariance_to_ros(bus: crate::geometry_msgs::msg::v1::PoseWithCovariance) -> ros_env::geometry_msgs::msg::PoseWithCovariance {
    ros_env::geometry_msgs::msg::PoseWithCovariance {
        pose: crate::ros2_bridge::mappers::geometry_msgs::pose::pose_to_ros(bus.pose.unwrap_or_default()),
        covariance: crate::ros2_bridge::mappers::convert::FromF64Seq::from_f64_seq(bus.covariance),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct GeometryMsgsPoseWithCovarianceMapper;

impl TypedTopicMapper for GeometryMsgsPoseWithCovarianceMapper {
    type Ros = ros_env::geometry_msgs::msg::PoseWithCovariance;
    type Bus = crate::geometry_msgs::msg::v1::PoseWithCovariance;

    fn type_name(&self) -> &'static str {
        "geometry_msgs/msg/PoseWithCovariance"
    }

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(pose_with_covariance_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(pose_with_covariance_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/geometry_msgs/pose_with_covariance.py
"""Generated mapper for `geometry_msgs/msg/PoseWithCovariance`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.geometry_msgs.pose import pose_to_bus, pose_to_ros

def pose_with_covariance_to_bus(msg):
    from robot_bus.geometry_msgs.msg.v1 import PoseWithCovariance as BusMsg

    bus = BusMsg()
    bus.pose.CopyFrom(pose_to_bus(msg.pose))
    bus.covariance.extend(list(msg.covariance))
    return bus


def pose_with_covariance_to_ros(bus):
    from geometry_msgs.msg import PoseWithCovariance as RosMsg

    out = RosMsg()
    out.pose = pose_to_ros(bus.pose)
    out.covariance = list(bus.covariance)
    return out


class GeometryMsgsPoseWithCovarianceMapper:
    def type_name(self) -> str:
        return "geometry_msgs/msg/PoseWithCovariance"

    def ros_msg_type(self):
        from geometry_msgs.msg import PoseWithCovariance as RosMsg

        return RosMsg

    def ros_to_bus(self, msg) -> bytes:
        return pose_with_covariance_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        from robot_bus.geometry_msgs.msg.v1 import PoseWithCovariance as BusMsg

        bus = BusMsg()
        bus.ParseFromString(payload)
        return pose_with_covariance_to_ros(bus)
